### PR TITLE
Refactor optimization

### PR DESCRIPTION
**Pull request by erikkjellgren**
- All wave functions now have two different optimization calls, run_wf_optimization_1step  and run_wf_optimization_2step
- Wave functions no longer take in num_spin_orbs  as an argument
- The circuit based wave function has changed name from WaveFunction  to WaveFunctionCircuit
- Changed c_trans  to c_mo  to make it more clear it is the mo coefficients.
- All wave functions now have self.thetas  for the active-space parameterization, i.e. self.ansatz_parameters  have been removed from the circuit based wave function.

### CODE REVIEW

File: slowquant/qiskit_interface/circuit_wavefunction.py [hunk 1]
```diff
@@ -142,67 +138,58 @@ def __init__(
             if idx // 2 not in self.active_unocc_idx:
                 self.active_unocc_idx.append(idx // 2)
         # Find non-redundant kappas
-        self.kappa = []
+        self._kappa = []
```
[the remark this answers is not in the document]
Comment (erikkjellgren) [reply slowquant/qiskit_interface/circuit_wavefunction.py:141]: Because the indicies might be needed outside of the wavefunction class, but if kappa is modified from the outside, bad things might happen xd

File: slowquant/qiskit_interface/util.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ def to_CBS_measurement(op: str, transpiled: None | list[QuantumCircuit] = None)
         transpiled: List of transpiled X and Y gate.
 
     Returns:
-        Pauli measuremnt quantum circuit.
```
Comment (ziechys) [inline slowquant/qiskit_interface/util.py:26]:
```suggestion
        Pauli measurement quantum circuit.
```

File: slowquant/qiskit_interface/optimizers.py [hunk 3]
```diff
@@ -1,169 +0,0 @@
...
-        self._param_names = param_names
-
-    def minimize(
-        self, f: Callable[[list[float]], float], x: list[float], jac=None  # pylint: disable=unused-argument
-    ) -> Result:
-        """Run minimization.
-
-        Args:
-            f: Function to be minimzed, can only take one argument.
-            x: Changable parameters of f.
```
Comment (ziechys) [inline slowquant/qiskit_interface/optimizers.py:76]:
```suggestion
            x: Changeable parameters of f.
```

File: slowquant/qiskit_interface/circuit_wavefunction.py [hunk 4]
```diff
@@ -1041,262 +947,173 @@ def print_progress_SPSA(
...
+        """
+        num_kappa = 0
+        if kappa_optimization:
+            num_kappa = len(self.kappa_idx)
+            self.kappa = parameters[:num_kappa]
+            self._move_cep()
+        if theta_optimization:
+            self.thetas = parameters[num_kappa:]
+        # Build operator
+        if theta_optimization:
```
Comment (ziechys) [inline slowquant/qiskit_interface/circuit_wavefunction.py:1049]:
```suggestion
```

File: slowquant/qiskit_interface/circuit_wavefunction.py [hunk 5]
```diff
@@ -1041,262 +947,173 @@ def print_progress_SPSA(
...
+            Electronic energy.
+        """
+        num_kappa = 0
+        if kappa_optimization:
+            num_kappa = len(self.kappa_idx)
+            self.kappa = parameters[:num_kappa]
+            self._move_cep()
+        if theta_optimization:
+            self.thetas = parameters[num_kappa:]
+        # Build operator
```
Comment (ziechys) [inline slowquant/qiskit_interface/circuit_wavefunction.py:1048]:
```suggestion
            # Build operator
```

File: slowquant/qiskit_interface/circuit_wavefunction.py [hunk 6]
```diff
@@ -1041,262 +947,173 @@ def print_progress_SPSA(
...
+        if theta_optimization:
+            H = hamiltonian_0i_0a(self.h_mo, self.g_mo, self.num_inactive_orbs, self.num_active_orbs)
+            H = H.get_folded_operator(self.num_inactive_orbs, self.num_active_orbs, self.num_virtual_orbs)
+            return self.QI.quantum_expectation_value(H)
+        rdms = ReducedDenstiyMatrix(
```
Comment (ziechys) [inline slowquant/qiskit_interface/circuit_wavefunction.py:1053]: Why is the energy only calcuated via RDM if theta_optimization = False?

File: slowquant/qiskit_interface/circuit_wavefunction.py [hunk 7]
```diff
@@ -1041,262 +947,173 @@ def print_progress_SPSA(
...
     quantum_interface: QuantumInterface,
     parameters: list[float],
     idx: int,
     R: int,
 ) -> list[float]:
-    r"""Get energy evaluations needed for the gradient calculation.
+    """Get energy evaluations needed for the gradient calculation.
 
     The gradient formula is defined for x=0,
     so x_shift is used to shift ensure we can get the energy in the point we actually want.
```
Comment (ziechys) [inline slowquant/qiskit_interface/circuit_wavefunction.py:1119]: Can you rephrase this sentence. Sounds strange

File: slowquant/unitary_coupled_cluster/optimizers.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+        f: Function to evaluate.
+        x: Parameters of f.
+        idx: Index of parameter to be changed.
+        R: Parameter to control how many points are needed.
+
+    Returns:
+        All needed function evaluations.
+    """
+    e_vals = []
+    x = x.copy()
```
Comment (ziechys) [inline slowquant/unitary_coupled_cluster/optimizers.py:258]: No x_shift here, compared to circuit_wavefunction.py? Is it absorbed in the equation?

File: slowquant/unitary_coupled_cluster/optimizers.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+            if fails == self.max_fail:
+                break
+            if self._callback is not None:
+                self._callback(x)
+        res.x = np.array(x_best)
+        res.fun = f_best
+        return res
+
+
+def get_energy_evals(
```
Comment (ziechys) [inline slowquant/unitary_coupled_cluster/optimizers.py:243]: In principle this function already exists in circuit_wave_function.py with f = quantum_expectation_value. Could be combined.

File: slowquant/unitary_coupled_cluster/optimizers.py [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+    f: Callable[[list[float]], float | list[float]], x: list[float], idx: int, R: int
+) -> list[float] | list[list[float]]:
+    """Evaluate the function in all points needed for the reconstrucing in Rotosolve.
```
Comment (ziechys) [inline slowquant/unitary_coupled_cluster/optimizers.py:246]:
```suggestion
    """Evaluate the function in all points needed for the reconstruction in Rotosolve.
```

File: slowquant/unitary_coupled_cluster/optimizers.py [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+        f_best = float(10**20)
+        x = list(x0).copy()
+        x_best = x.copy()
+        fails = 0
+        res = Result()
+        for _ in range(self.max_iterations):
+            for i, par_name in enumerate(self._param_names):
+                e_vals = get_energy_evals(f, x, i, self._R[par_name])
+                f_reconstructed = partial(reconstructed_f, energy_vals=e_vals, R=self._R[par_name])
+                values = f_reconstructed(np.linspace(-np.pi, np.pi, int(1e4)))
```
Comment (ziechys) [inline slowquant/unitary_coupled_cluster/optimizers.py:210]: Could you maybe just add a few comments here. This would be really nice. Like 1 line comments for each of these important steps

File: slowquant/qiskit_interface/circuit_wavefunction.py (hunk 6 above)
Comment (erikkjellgren) [reply slowquant/qiskit_interface/circuit_wavefunction.py:1052]: Because RDM construction is the same price or more expensive than calculating the Hamiltonian (some of the RDM terms might not be needed for the energy because the integrals are zero).

But if theta_optimization is false, then the RDM's calculated to get the energy will also be used to get the gradients for the orbital optimization, hence it is worth it then.

File: slowquant/qiskit_interface/circuit_wavefunction.py [hunk 12]
```diff
@@ -1041,262 +947,173 @@ def print_progress_SPSA(
...
-
-
-def get_energy_evals_for_grad(
+        Returns:
+            Electronic gradient.
+        """
+        num_kappa = 0
+        gradient = np.zeros(len(parameters))
+        num_kappa = 0
+        if kappa_optimization:
```
[the remark this answers is not in the document]
Comment (erikkjellgren) [reply slowquant/qiskit_interface/circuit_wavefunction.py:1077]: Because I do not assume any specific implementation details of the underlying optimizers.
In principle a gradient descent does not need to evaluate the energy but could just call the gradient many times.
In this case the current-expansion-point also need to be moved.

File: slowquant/unitary_coupled_cluster/optimizers.py (hunk 11 above)
Comment (erikkjellgren) [reply slowquant/unitary_coupled_cluster/optimizers.py:213]: I have granted you some comments ...

File: slowquant/qiskit_interface/circuit_wavefunction.py [hunk 13]
```diff
@@ -142,96 +138,78 @@ def __init__(
...
-                for kappa_val, (p, q) in zip(self.kappa, self.kappa_idx):
-                    kappa_mat[p, q] = kappa_val
-                    kappa_mat[q, p] = -kappa_val
-        if len(self.kappa_redundant) != 0:
-            if np.max(np.abs(self.kappa_redundant)) > 0.0:
-                for kappa_val, (p, q) in zip(self.kappa_redundant, self.kappa_redundant_idx):
-                    kappa_mat[p, q] = kappa_val
-                    kappa_mat[q, p] = -kappa_val
-        return np.matmul(self._c_orthonormal, scipy.linalg.expm(-kappa_mat))
+            if np.max(np.abs(np.array(self.kappa) - np.array(self._kappa_old))) > 0.0:
```
Comment (ziechys) [inline slowquant/qiskit_interface/circuit_wavefunction.py:208]: A comment to why we build the kappa matrix as difference values maybe? (sth about expansion point changes)
#readibility

File: slowquant/qiskit_interface/circuit_wavefunction.py (hunk 6 above)
Comment (ziechys) [reply slowquant/qiskit_interface/circuit_wavefunction.py:1052]: Ok. makes sense. How about a one-liner comment explaining this

File: slowquant/qiskit_interface/circuit_wavefunction.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ziechys) [reply slowquant/qiskit_interface/circuit_wavefunction.py:1077]: Ok. And otherwise it would just mean that kappa = kappa_old and the expansion point is not actually moved when calling _move_cep?

File: slowquant/unitary_coupled_cluster/ucc_wavefunction.py [hunk 14]
```diff
@@ -26,16 +26,16 @@
     get_indexing,
 )
 from slowquant.unitary_coupled_cluster.operators import Epq, hamiltonian_0i_0a
+from slowquant.unitary_coupled_cluster.optimizers import Optimizers
 from slowquant.unitary_coupled_cluster.util import UccStructure
 
 
 class WaveFunctionUCC:
```
Comment (ziechys) [inline slowquant/unitary_coupled_cluster/ucc_wavefunction.py:33]: Harmonize (new) comments to be equal in all wave function classes

File: slowquant/unitary_coupled_cluster/sa_ups_wavefunction.py [hunk 15]
```diff
@@ -30,16 +30,16 @@
     hamiltonian_0i_0a,
     one_elec_op_0i_0a,
 )
+from slowquant.unitary_coupled_cluster.optimizers import Optimizers
 from slowquant.unitary_coupled_cluster.util import UpsStructure
 
 
 class WaveFunctionSAUPS:
```
Comment (ziechys) [inline slowquant/unitary_coupled_cluster/sa_ups_wavefunction.py:37]: Harmonize (new) comments to be equal in all wave function classes

File: slowquant/unitary_coupled_cluster/ups_wavefunction.py [hunk 16]
```diff
@@ -27,16 +27,16 @@
     propagate_unitary,
 )
 from slowquant.unitary_coupled_cluster.operators import Epq, hamiltonian_0i_0a
+from slowquant.unitary_coupled_cluster.optimizers import Optimizers
 from slowquant.unitary_coupled_cluster.util import UpsStructure
 
 
 class WaveFunctionUPS:
```
Comment (ziechys) [inline slowquant/unitary_coupled_cluster/ups_wavefunction.py:34]: Harmonize (new) comments to be equal in all wave function classes

**Review by ziechys — APPROVED** [review]

File: slowquant/qiskit_interface/circuit_wavefunction.py [hunk 17]
```diff
@@ -278,6 +256,15 @@ def ansatz_parameters(self, parameters: list[float]) -> None:
         self._energy_elec = None
         self.QI.parameters = parameters
 
+    def _move_cep(self) -> None:
+        """Move current expansion point."""
```
[the remark this answers is not in the document]
Comment (erikkjellgren) [reply slowquant/qiskit_interface/circuit_wavefunction.py:260]: You are right, this is just unnecessary complexity.

**Review by ziechys — APPROVED** [review]
Beautiful new comments :)
